test_rmw_implementation/test/test_client.cpp

```diff
@@ -0,0 +1,186 @@
+  native_qos_profile.avoid_ros_namespace_conventions = true;
+  rmw_client_t * client =
+    rmw_create_client(node, ts, service_name, &native_qos_profile);
+  ASSERT_NE(nullptr, client) << rmw_get_error_string().str;
+  rmw_ret_t ret = rmw_destroy_client(node, client);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestClient, RMW_IMPLEMENTATION), create_with_bad_arguments) {
+  constexpr char service_name[] = "/test";
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+
+  rmw_client_t * client =
+    rmw_create_client(nullptr, ts, service_name, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  client = rmw_create_client(node, nullptr, service_name, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  client = rmw_create_client(node, ts, service_name, &rmw_qos_profile_default);
+  node->implementation_identifier = implementation_identifier;
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  client = rmw_create_client(node, ts, nullptr, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  client = rmw_create_client(node, ts, "", &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  constexpr char service_name_with_spaces[] = "/foo bar";
+  client = rmw_create_client(node, ts, service_name_with_spaces, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  constexpr char relative_service_name[] = "foo";
+  client = rmw_create_client(node, ts, relative_service_name, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  client = rmw_create_client(node, ts, service_name, nullptr);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  client = rmw_create_client(node, ts, service_name, &rmw_qos_profile_unknown);
+  EXPECT_EQ(nullptr, client);
+  rmw_reset_error();
+
+  // Creating and destroying a client still succeeds.
+  client = rmw_create_client(node, ts, service_name, &rmw_qos_profile_default);
+  ASSERT_NE(nullptr, client) << rmw_get_error_string().str;
+  rmw_ret_t ret = rmw_destroy_client(node, client);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestClient, RMW_IMPLEMENTATION), destroy_with_bad_arguments) {
+  constexpr char service_name[] = "/test";
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+  rmw_client_t * client =
+    rmw_create_client(node, ts, service_name, &rmw_qos_profile_default);
+  ASSERT_NE(nullptr, client) << rmw_get_error_string().str;
+
+  // Destroying client with invalid arguments fails.
+  rmw_ret_t ret = rmw_destroy_client(nullptr, client);
```

Comment:
If this unintentionally modifies client, it could impact the other checks below. It might be better to recreate client each time you try to destroy it, or check it's valid each time

Reply:
Same as above.